File: cpp/searching/binary_search.cpp
#include <vector>

int binarySearch(std::vector<int> &arr, int target)
{
    int first{0};
    int last = arr.size() - 1;

    do
    {
        int mid{first + (last - first) / 2};

        if (arr[mid] == target)
            return mid;
        else if (arr[mid] > target)
            last = mid - 1;
        else
            first = mid + 1;

    } while (first <= last);

    return -1;
}

File: python/searching/binary_search.py
# target array must be sorted
def binary_search(array, target):
    first = 0
    last = len(array) - 1

    while first <= last:
        mid = (first + last) // 2
        if array[mid] == target:
            return mid
        elif array[mid] > target:
            last = mid - 1
        elif array[mid] < target:
            first = mid + 1

    return -1
